File: leetcode_75/1_Array_String/python/605.py
from typing import List


class Solution:
    def canPlaceFlowers(self, flowerbed: List[int], n: int) -> bool:
        counter = 0

        if len(flowerbed) == 1 and flowerbed[0] == 0:
            counter += 1

        if len(flowerbed) >= 2:
            if not flowerbed[0] and not flowerbed[1]:
                flowerbed[0] = 1
                counter += 1
        for i in range(2, len(flowerbed)):
            if all(not x for x in flowerbed[i - 2: i + 1]):
                flowerbed[i - 1] = 1
                counter += 1
            if counter == n:
                break

        if len(flowerbed) >= 2 and not flowerbed[-2] and not flowerbed[-1]:
            counter += 1

        if counter >= n:
            return True
        return False


solution = Solution()
print(solution.canPlaceFlowers([1, 0, 0, 0, 1, 0, 0], n=2))

File: leetcode_75/Array_String/cpp/605.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution
{
public:
    bool canPlaceFlowers(vector<int> &flowerbed, int n)
    {
        int counter = 0;
        int size = flowerbed.size();

        if (size == 1 && flowerbed[0] == 0)
        {
            counter++;
        }

        if (size >= 2 && flowerbed[0] + flowerbed[1] == 0)
        {
            flowerbed[0] = 1;
            counter++;
        }

        for (int i = 2; i < size; ++i)
        {
            if (flowerbed[i - 2] + flowerbed[i - 1] + flowerbed[i] == 0)
            {
                flowerbed[i - 1] = 1;
                counter++;
            }
            if (counter == n)
            {
                break;
            }
        }

        if (size >= 2 && flowerbed[size - 2] + flowerbed[size - 1] == 0)
        {
            counter++;
        }

        return counter >= n ? true : false;
    }
};

int main()
{
    std::vector<int> flowerbed = {1, 0, 0, 0, 1};
    Solution solution;
    int n = 1;
    bool canPlace = solution.canPlaceFlowers(flowerbed, n);

    std::cout << "Can place flowers: " << canPlace << std::endl;
    return 0;
}